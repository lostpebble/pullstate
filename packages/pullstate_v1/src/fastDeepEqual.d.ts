declare module "fast-deep-equal/es6" {
  const equal: (a: any, b: any) => boolean;
  export = equal;
}
