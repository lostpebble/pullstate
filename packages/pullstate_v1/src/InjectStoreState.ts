import React from "react";
import { Store } from "./Store";
import { useStoreState } from "./useStoreState";

export interface IPropsInjectStoreState<S extends object = any, SS extends any = any> {
  store: Store<S>;
  on?: (state: S) => SS;
  children: (output: SS) => React.ReactElement;
}

export function InjectStoreState<S extends object = any, SS = any>({
  store,
  on = s => s as any,
  children,
}: IPropsInjectStoreState<S, SS>): React.ReactElement {
  const state: SS = useStoreState(store, on);
  return children(state);
}
