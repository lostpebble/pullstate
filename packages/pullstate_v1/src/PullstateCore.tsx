import React, { useContext } from "react";
import { Store, TUpdateFunction } from "./Store";
import { clientAsyncCache, createAsyncAction, createAsyncActionDirect } from "./async";
import {
  IAsyncActionRunOptions,
  ICreateAsyncActionOptions,
  IOCreateAsyncActionOutput,
  IPullstateAsyncActionOrdState,
  IPullstateAsyncCache,
  IPullstateAsyncResultState,
  TPullstateAsyncAction,
  TPullstateAsyncRunResponse
} from "./async-types";

export interface IPullstateAllStores {
  [storeName: string]: Store<any>;
}

export const PullstateContext = React.createContext<PullstateInstance<any> | null>(null);

export const PullstateProvider = <T extends IPullstateAllStores>(
  {
    instance,
    children
  }: {
    instance: PullstateInstance<T>;
    children?: any;
  }) => {
  return <PullstateContext.Provider value={instance}>{children}</PullstateContext.Provider>;
};

let singleton: PullstateSingleton<any> | null = null;

export const clientStores: {
  internalClientStores: true;
  stores: IPullstateAllStores;
  loaded: boolean;
} = {
  internalClientStores: true,
  loaded: false,
  stores: {}
};

export type TMultiStoreAction<P extends PullstateSingleton<S>,
  S extends IPullstateAllStores = P extends PullstateSingleton<infer ST> ? ST : any> = (update: TMultiStoreUpdateMap<S>) => void;

interface IPullstateSingletonOptions {
  asyncActions?: {
    defaultCachingSeconds?: number;
  };
}

export class PullstateSingleton<S extends IPullstateAllStores = IPullstateAllStores> {
  // private readonly originStores: S = {} as S;
  // private updatedStoresInAct = new Set<string>();
  // private actUpdateMap: TMultiStoreUpdateMap<S> | undefined;
  options: IPullstateSingletonOptions = {};

  constructor(allStores: S, options: IPullstateSingletonOptions = {}) {
    if (singleton !== null) {
      console.error(
        `Pullstate: createPullstate() - Should not be creating the core Pullstate class more than once! In order to re-use pull state, you need to call instantiate() on your already created object.`
      );
    }

    singleton = this;
    // this.originStores = allStores;
    clientStores.stores = allStores;
    clientStores.loaded = true;
    this.options = options;
  }

  instantiate(
    {
      hydrateSnapshot,
      ssr = false,
      customContext
    }: { hydrateSnapshot?: IPullstateSnapshot; ssr?: boolean, customContext?: any } = {}): PullstateInstance<S> {
    if (!ssr) {
      const instantiated = new PullstateInstance<S>(clientStores.stores as S, false, customContext);

      if (hydrateSnapshot != null) {
        instantiated.hydrateFromSnapshot(hydrateSnapshot);
      }

      instantiated.instantiateReactions();
      return instantiated as PullstateInstance<S>;
    }

    const newStores: IPullstateAllStores = {};

    for (const storeName of Object.keys(clientStores.stores)) {
      if (hydrateSnapshot == null) {
        newStores[storeName] = new Store(clientStores.stores[storeName]._getInitialState());
      } else if (hydrateSnapshot.hasOwnProperty(storeName)) {
        newStores[storeName] = new Store(hydrateSnapshot.allState[storeName]);
      } else {
        newStores[storeName] = new Store(clientStores.stores[storeName]._getInitialState());
        console.warn(
          `Pullstate (instantiate): store [${storeName}] didn't hydrate any state (data was non-existent on hydration object)`
        );
      }

      newStores[storeName]._setInternalOptions({
        ssr,
        reactionCreators: clientStores.stores[storeName]._getReactionCreators()
      });
    }

    return new PullstateInstance(newStores as S, true, customContext);
  }

  useStores(): S {
    // return useContext(PullstateContext)!.stores as S;
    return useStores<S>();
  }

  useInstance(): PullstateInstance<S> {
    return useInstance<S>();
  }

  /*actionSetup(): {
    action: (update: TMultiStoreAction<PullstateSingleton<S>, S>) => TMultiStoreAction<PullstateSingleton<S>, S>;
    act: (action: TMultiStoreAction<PullstateSingleton<S>, S>) => void;
    // act: (action: (update: TMultiStoreUpdateMap<S>) => void) => void;
  } {
    const actUpdateMap = {} as TMultiStoreUpdateMap<S>;
    const updatedStores = new Set<string>();

    for (const store of Object.keys(clientStores.stores)) {
      actUpdateMap[store as keyof S] = (updater) => {
        updatedStores.add(store);
        clientStores.stores[store].batch(updater);
      };
    }

    const action: (
      update: TMultiStoreAction<PullstateSingleton<S>, S>
    ) => TMultiStoreAction<PullstateSingleton<S>, S> = (action) => action;
    const act = (action: TMultiStoreAction<PullstateSingleton<S>, S>): void => {
      updatedStores.clear();
      action(actUpdateMap);
      for (const store of updatedStores) {
        clientStores.stores[store].flushBatch(true);
      }
    };

    return {
      action,
      act,
    };
  }*/

  createAsyncActionDirect<A extends any = any, R extends any = any, N extends any = any>(
    action: (args: A) => Promise<R>,
    options: ICreateAsyncActionOptions<A, R, string, N, S> = {}
  ): IOCreateAsyncActionOutput<A, R, string, N, S> {
    return createAsyncActionDirect(action, options);
    // return createAsyncAction<A, R, string, S>(async (args: A) => {
    //   return successResult(await action(args));
    // }, options);
  }

  createAsyncAction<A = any, R = any, T extends string = string, N extends any = any>(
    action: TPullstateAsyncAction<A, R, T, N, S>,
    // options: Omit<ICreateAsyncActionOptions<A, R, T, S>, "clientStores"> = {}
    options: ICreateAsyncActionOptions<A, R, T, N, S> = {}
  ): IOCreateAsyncActionOutput<A, R, T, N, S> {
    // options.clientStores = this.originStores;
    if (this.options.asyncActions?.defaultCachingSeconds && !options.cacheBreakHook) {
      options.cacheBreakHook = (inputs) =>
        inputs.timeCached < Date.now() - this.options.asyncActions!.defaultCachingSeconds! * 1000;
    }

    return createAsyncAction<A, R, T, N, S>(action, options);
  }
}

type TMultiStoreUpdateMap<S extends IPullstateAllStores> = {
  [K in keyof S]: (updater: TUpdateFunction<S[K] extends Store<infer T> ? T : any>) => void;
};

interface IPullstateSnapshot {
  allState: { [storeName: string]: any };
  asyncResults: IPullstateAsyncResultState;
  asyncActionOrd: IPullstateAsyncActionOrdState;
}

export interface IPullstateInstanceConsumable<T extends IPullstateAllStores = IPullstateAllStores> {
  stores: T;

  hasAsyncStateToResolve(): boolean;

  resolveAsyncState(): Promise<void>;

  getPullstateSnapshot(): IPullstateSnapshot;

  hydrateFromSnapshot(snapshot: IPullstateSnapshot): void;

  runAsyncAction<A, R, X extends string, N>(
    asyncAction: IOCreateAsyncActionOutput<A, R, X, N, T>,
    args?: A,
    runOptions?: Pick<IAsyncActionRunOptions<A, R, X, N, T>, "ignoreShortCircuit" | "respectCache">
  ): TPullstateAsyncRunResponse<R, X, N>;
}

class PullstateInstance<T extends IPullstateAllStores = IPullstateAllStores>
  implements IPullstateInstanceConsumable<T> {
  private _ssr: boolean = false;
  private _customContext: any;
  private readonly _stores: T = {} as T;
  _asyncCache: IPullstateAsyncCache = {
    listeners: {},
    results: {},
    actions: {},
    actionOrd: {}
  };

  constructor(allStores: T, ssr: boolean, customContext: any) {
    this._stores = allStores;
    this._ssr = ssr;
    this._customContext = customContext;
    /*if (!ssr) {
      // console.log(`Instantiating Stores`, allStores);
      clientStores.stores = allStores;
      clientStores.loaded = true;
    }*/
  }

  private getAllUnresolvedAsyncActions(): Array<Promise<any>> {
    return Object.keys(this._asyncCache.actions).map((key) => this._asyncCache.actions[key]());
  }

  instantiateReactions() {
    for (const storeName of Object.keys(this._stores)) {
      this._stores[storeName]._instantiateReactions();
    }
  }

  getPullstateSnapshot(): IPullstateSnapshot {
    const allState = {} as IPullstateSnapshot["allState"];

    for (const storeName of Object.keys(this._stores)) {
      allState[storeName] = this._stores[storeName].getRawState();
    }

    return { allState, asyncResults: this._asyncCache.results, asyncActionOrd: this._asyncCache.actionOrd };
  }

  async resolveAsyncState() {
    const promises = this.getAllUnresolvedAsyncActions();
    await Promise.all(promises);
  }

  hasAsyncStateToResolve(): boolean {
    return Object.keys(this._asyncCache.actions).length > 0;
  }

  get stores(): T {
    return this._stores;
  }

  get customContext(): any {
    return this._customContext;
  }

  async runAsyncAction<A, R, X extends string, N>(
    asyncAction: IOCreateAsyncActionOutput<A, R, X, N, T>,
    args: A = {} as A,
    runOptions: Pick<IAsyncActionRunOptions<A, R, X, N, T>, "ignoreShortCircuit" | "respectCache"> = {}
  ): TPullstateAsyncRunResponse<R, X, N> {
    if (this._ssr) {
      (runOptions as IAsyncActionRunOptions<A, R, X, N, T>)._asyncCache = this._asyncCache;
      (runOptions as IAsyncActionRunOptions<A, R, X, N, T>)._stores = this._stores;
      (runOptions as IAsyncActionRunOptions<A, R, X, N, T>)._customContext = this._customContext;
    }

    return await asyncAction.run(args, runOptions);
  }

  hydrateFromSnapshot(snapshot: IPullstateSnapshot) {
    for (const storeName of Object.keys(this._stores)) {
      if (snapshot.allState.hasOwnProperty(storeName)) {
        this._stores[storeName]._updateStateWithoutReaction(snapshot.allState[storeName]);
      } else {
        console.warn(`${storeName} didn't hydrate any state (data was non-existent on hydration object)`);
      }
    }

    clientAsyncCache.results = snapshot.asyncResults || {};
    clientAsyncCache.actionOrd = snapshot.asyncActionOrd || {};
  }
}

export function createPullstateCore<T extends IPullstateAllStores = IPullstateAllStores>(
  allStores: T = {} as T,
  options: IPullstateSingletonOptions = {}
) {
  return new PullstateSingleton<T>(allStores, options);
}

export function useStores<T extends IPullstateAllStores = {}>() {
  return useContext(PullstateContext)!.stores as T;
}

export function useInstance<T extends IPullstateAllStores = IPullstateAllStores>(): PullstateInstance<T> {
  return useContext(PullstateContext)! as PullstateInstance<T>;
}
