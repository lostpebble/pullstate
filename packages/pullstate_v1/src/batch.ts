import { globalClientState } from "./globalClientState";

interface IBatchState {
  uiBatchFunction: ((updates: () => void) => void);
}

const batchState: Partial<IBatchState> = {};

export function setupBatch({ uiBatchFunction }: IBatchState) {
  batchState.uiBatchFunction = uiBatchFunction;
}

export function batch(runUpdates: () => void) {
  if (globalClientState.batching) {
    throw new Error("Pullstate: Can't enact two batch() update functions at the same time-\n" +
      "make sure you are not running a batch() inside of a batch() by mistake.");
  }

  globalClientState.batching = true;

  try {
    runUpdates();
  } finally {
    if (batchState.uiBatchFunction) {
      batchState.uiBatchFunction(() => {
        Object.values(globalClientState.flushStores).forEach(store => store.flushBatch(true));
      });
    } else {
      Object.values(globalClientState.flushStores).forEach(store => store.flushBatch(true));
    }
    globalClientState.flushStores = {};
    globalClientState.batching = false;
  }
}
