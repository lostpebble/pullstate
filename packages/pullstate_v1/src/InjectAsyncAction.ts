import React from "react";
import {
  IAsyncActionBeckonOptions,
  IAsyncActionWatchOptions,
  IOCreateAsyncActionOutput,
  TPullstateAsyncBeckonResponse,
  TPullstateAsyncWatchResponse
} from "./async-types";
import { IPullstateAllStores } from "./PullstateCore";

export enum EAsyncActionInjectType {
  WATCH = "watch",
  BECKON = "beckon",
}

interface IPropsInjectAsyncActionBase<A, R, T extends string, N> {
  action: IOCreateAsyncActionOutput<A, R, T, N>;
  args?: A;
}

export interface IPropsInjectAsyncActionBeckon<A = any, R = any, T extends string = string, N = any, S extends IPullstateAllStores = IPullstateAllStores>
  extends IPropsInjectAsyncActionBase<A, R, T, N> {
  type: EAsyncActionInjectType.BECKON;
  options?: IAsyncActionBeckonOptions<A, R, T, N, S>;
  children: (response: TPullstateAsyncBeckonResponse<R, T>) => React.ReactElement;
}

export interface IPropsInjectAsyncActionWatch<A = any, R = any, T extends string = string, N = any, S extends IPullstateAllStores = IPullstateAllStores>
  extends IPropsInjectAsyncActionBase<A, R, T, N> {
  type: EAsyncActionInjectType.WATCH;
  children: (response: TPullstateAsyncWatchResponse<R, T, N>) => React.ReactElement;
  options?: IAsyncActionWatchOptions<A, R, T, N, S>;
}

export type TInjectAsyncActionProps = IPropsInjectAsyncActionBeckon | IPropsInjectAsyncActionWatch;

export function InjectAsyncAction(
  props: TInjectAsyncActionProps
): React.ReactElement {
  if (props.type === EAsyncActionInjectType.BECKON) {
    const response = props.action.useBeckon(props.args, props.options);
    return props.children(response);
  }

  const response = props.action.useWatch(props.args, props.options);
  return props.children(response);
}
