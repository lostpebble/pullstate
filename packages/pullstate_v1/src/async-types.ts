import { IPullstateAllStores } from "./PullstateCore";
import { TUpdateFunction } from "./Store";

type TPullstateAsyncUpdateListener = () => void;

// [ started, finished, result, updating, timeCached ]
export type TPullstateAsyncWatchResponse<R = any, T extends string = string, N = any> = [
  boolean,
  boolean,
  TAsyncActionResult<R, T, N>,
  boolean,
  number
];

// export type TPullstateAsync

// [ started, finished, result, updating, postActionResult ]
// export type TPullstateAsyncResponseCacheFull<R, T extends string> = [
//   boolean,
//   boolean,
//   TAsyncActionResult<R, T>,
//   boolean,
//   TAsyncActionResult<R, T> | true | null
// ];

// [finished, result, updating]
export type TPullstateAsyncBeckonResponse<R = any, T extends string = string, N = any> = [
  boolean,
  TAsyncActionResult<R, T, N>,
  boolean
];
// [result]
export type TPullstateAsyncRunResponse<R = any, T extends string = string, N = any> = Promise<TAsyncActionResult<R, T, N>>;

export interface IPullstateAsyncResultState {
  [key: string]: TPullstateAsyncWatchResponse<any, string, any>;
}

export interface IPullstateAsyncActionOrdState {
  [key: string]: number;
}

export enum EAsyncEndTags {
  THREW_ERROR = "THREW_ERROR",
  RETURNED_ERROR = "RETURNED_ERROR",
  UNFINISHED = "UNFINISHED",
  DORMANT = "DORMANT",
}

interface IAsyncActionResultBase<T extends string> {
  message: string;
  tags: (EAsyncEndTags | T)[];
}

export interface IAsyncActionResultPositive<R, T extends string> extends IAsyncActionResultBase<T> {
  error: false;
  payload: R;
  errorPayload: null;
}

export interface IAsyncActionResultNegative<T extends string, N = unknown> extends IAsyncActionResultBase<T> {
  error: true;
  errorPayload: N;
  payload: null;
}

export type TAsyncActionResult<R, T extends string, N> =
  IAsyncActionResultPositive<R, T>
  | IAsyncActionResultNegative<T, N>;

// Order of new hook functions:

// shortCircuitHook = ({ args, stores }) => cachable response | false     - happens only on uncached action
// cacheBreakHook = ({ args, stores, result }) => true | false            - happens only on cached action
// postActionHook = ({ args, result, stores }) => void | new result       - happens on all actions, after the async / short circuit has resolved
// ----> postActionHook potentially needs a mechanism which allows it to run only once per new key change (another layer caching of some sorts expiring on key change)

export type TPullstateAsyncShortCircuitHook<A, R, T extends string, N, S extends IPullstateAllStores> = (inputs: {
  args: A;
  stores: S;
}) => TAsyncActionResult<R, T, N> | false;

export type TPullstateAsyncCacheBreakHook<A, R, T extends string, N, S extends IPullstateAllStores> = (inputs: {
  args: A;
  result: TAsyncActionResult<R, T, N>;
  stores: S;
  timeCached: number;
}) => boolean;

export enum EPostActionContext {
  WATCH_HIT_CACHE = "WATCH_HIT_CACHE",
  BECKON_HIT_CACHE = "BECKON_HIT_CACHE",
  RUN_HIT_CACHE = "RUN_HIT_CACHE",
  READ_HIT_CACHE = "READ_HIT_CACHE",
  READ_RUN = "READ_RUN",
  SHORT_CIRCUIT = "SHORT_CIRCUIT",
  DIRECT_RUN = "DIRECT_RUN",
  BECKON_RUN = "BECKON_RUN",
  CACHE_UPDATE = "CACHE_UPDATE",
}

export type TPullstateAsyncPostActionHook<A, R, T extends string, N, S extends IPullstateAllStores> = (inputs: {
  args: A;
  result: TAsyncActionResult<R, T, N>;
  stores: S;
  context: EPostActionContext;
}) => void;

export interface IAsyncActionReadOptions<A, R, T extends string, N, S extends IPullstateAllStores> {
  postActionEnabled?: boolean;
  cacheBreakEnabled?: boolean;
  key?: string;
  cacheBreak?: boolean | number | TPullstateAsyncCacheBreakHook<A, R, T, N, S>
}

export interface IAsyncActionBeckonOptions<A, R, T extends string, N, S extends IPullstateAllStores> extends IAsyncActionReadOptions<A, R, T, N, S> {
  ssr?: boolean;
  holdPrevious?: boolean;
  dormant?: boolean;
}

export interface IAsyncActionWatchOptions<A, R, T extends string, N, S extends IPullstateAllStores> extends IAsyncActionBeckonOptions<A, R, T, N, S> {
  initiate?: boolean;
}

export interface IAsyncActionUseOptions<A, R, T extends string, N, S extends IPullstateAllStores> extends IAsyncActionWatchOptions<A, R, T, N, S> {
  onSuccess?: (result: R, args: A) => void;
}

export interface IAsyncActionUseDeferOptions<A, R, T extends string, N, S extends IPullstateAllStores> extends Omit<IAsyncActionReadOptions<A, R, T, N, S>, "key"> {
  key?: string;
  holdPrevious?: boolean;
  onSuccess?: (result: R, args: A) => void;
  clearOnSuccess?: boolean;
}

export interface IAsyncActionRunOptions<A, R, T extends string, N, S extends IPullstateAllStores> {
  treatAsUpdate?: boolean;
  ignoreShortCircuit?: boolean;
  respectCache?: boolean;
  key?: string;
  cacheBreak?: boolean | number | TPullstateAsyncCacheBreakHook<A, R, T, N, S>
  _asyncCache?: IPullstateAsyncCache;
  _stores?: S;
  _customContext?: any;
}

export interface IAsyncActionGetCachedOptions<A, R, T extends string, N, S extends IPullstateAllStores> {
  checkCacheBreak?: boolean;
  cacheBreak?: boolean | number | TPullstateAsyncCacheBreakHook<A, R, T, N, S>;
  key?: string;
}

export interface IGetCachedResponse<R, T extends string, N = any> {
  started: boolean;
  finished: boolean;
  result: TAsyncActionResult<R, T, N>;
  updating: boolean;
  existed: boolean;
  cacheBreakable: boolean;
  timeCached: number;
}

export interface IAsyncClearCacheOptions {
  notify?: boolean;
}

export interface IAsyncActionSetOrClearCachedValueOptions extends IAsyncClearCacheOptions {
  key?: string;
}

export interface IAsyncActionUpdateCachedOptions extends IAsyncActionSetOrClearCachedValueOptions {
  resetTimeCached?: boolean;
  runPostActionHook?: boolean;
}

export type TAsyncActionUse<A, R, T extends string, N, S extends IPullstateAllStores> = (
  args?: A,
  options?: IAsyncActionUseOptions<A, R, T, N, S>
) => TUseResponse<A, R, T, N, S>;

export type TAsyncActionUseDefer<A, R, T extends string, N, S extends IPullstateAllStores> = (
  options?: IAsyncActionUseDeferOptions<A, R, T, N, S>
) => TUseDeferResponse<A, R, T, N, S>;

export type TAsyncActionBeckon<A, R, T extends string, N, S extends IPullstateAllStores> = (
  args?: A,
  options?: IAsyncActionBeckonOptions<A, R, T, N, S>
) => TPullstateAsyncBeckonResponse<R, T, N>;

export type TAsyncActionWatch<A, R, T extends string, N, S extends IPullstateAllStores> = (
  args?: A,
  options?: IAsyncActionWatchOptions<A, R, T, N, S>
) => TPullstateAsyncWatchResponse<R, T, N>;

export type TAsyncActionRun<A, R, T extends string, N, S extends IPullstateAllStores> = (
  args?: A,
  options?: IAsyncActionRunOptions<A, R, T, N, S>
) => TPullstateAsyncRunResponse<R, T, N>;

export type TAsyncActionClearCache<A> = (args?: A, options?: IAsyncActionSetOrClearCachedValueOptions) => void;

export type TAsyncActionClearAllCache = (options?: IAsyncClearCacheOptions) => void;

export type TAsyncActionClearAllUnwatchedCache = (options?: IAsyncClearCacheOptions) => void;

export type TAsyncActionGetCached<A, R, T extends string, N, S extends IPullstateAllStores> = (
  args?: A,
  options?: IAsyncActionGetCachedOptions<A, R, T, N, S>
) => IGetCachedResponse<R, T, N>;

export type TAsyncActionSetCached<A, R, T extends string, N> = (
  args: A,
  result: TAsyncActionResult<R, T, N>,
  options?: IAsyncActionSetOrClearCachedValueOptions
) => void;

export type TAsyncActionSetCachedPayload<A, R> = (args: A, payload: R, options?: IAsyncActionSetOrClearCachedValueOptions) => void;

export type TAsyncActionUpdateCached<A, R> = (
  args: A,
  updater: TUpdateFunction<R>,
  options?: IAsyncActionUpdateCachedOptions
) => void;
export type TAsyncActionRead<A, R, T extends string, N, S extends IPullstateAllStores> = (args?: A, options?: IAsyncActionReadOptions<A, R, T, N, S>) => R;

export type TAsyncActionDelayedRun<A, R, T extends string, N, S extends IPullstateAllStores> = (
  args: A,
  options: IAsyncActionRunOptions<A, R, T, N, S> & { delay: number; clearOldRun?: boolean; immediateIfCached?: boolean }
) => () => void;

export interface IOCreateAsyncActionOutput<A = any, R = any, T extends string = string, N = any, S extends IPullstateAllStores = IPullstateAllStores> {
  use: TAsyncActionUse<A, R, T, N, S>;
  useDefer: TAsyncActionUseDefer<A, R, T, N, S>;
  read: TAsyncActionRead<A, R, T, N, S>;
  useBeckon: TAsyncActionBeckon<A, R, T, N, S>;
  useWatch: TAsyncActionWatch<A, R, T, N, S>;
  run: TAsyncActionRun<A, R, T, N, S>;
  delayedRun: TAsyncActionDelayedRun<A, R, T, N, S>;
  getCached: TAsyncActionGetCached<A, R, T, N, S>;
  setCached: TAsyncActionSetCached<A, R, T, N>;
  setCachedPayload: TAsyncActionSetCachedPayload<A, R>;
  updateCached: TAsyncActionUpdateCached<A, R>;
  clearCache: TAsyncActionClearCache<A>;
  clearAllCache: TAsyncActionClearAllCache;
  clearAllUnwatchedCache: TAsyncActionClearAllUnwatchedCache;
}

export interface IPullstateAsyncCache {
  results: IPullstateAsyncResultState;
  listeners: {
    [key: string]: {
      [watchId: string]: TPullstateAsyncUpdateListener;
    };
  };
  actions: {
    [key: string]: () => Promise<TAsyncActionResult<any, string, any>>;
  };
  actionOrd: IPullstateAsyncActionOrdState;
}

export type TPullstateAsyncAction<A, R, T extends string, N, S extends IPullstateAllStores> = (
  args: A,
  stores: S,
  customContext: any
) => Promise<TAsyncActionResult<R, T, N>>;

export interface ICreateAsyncActionOptions<A, R, T extends string, N, S extends IPullstateAllStores> {
  forceContext?: boolean;
  // clientStores?: S;
  shortCircuitHook?: TPullstateAsyncShortCircuitHook<A, R, T, N, S>;
  cacheBreakHook?: TPullstateAsyncCacheBreakHook<A, R, T, N, S>;
  postActionHook?: TPullstateAsyncPostActionHook<A, R, T, N, S>;
  subsetKey?: (args: A) => any;
  actionId?: string | number;
}

// action.use() types

export interface IUseDebouncedExecutionOptions<A, R, T extends string, N, S extends IPullstateAllStores> {
  validInput?: (args: A) => boolean;
  equality?: ((argsPrev: A, argsNew: A) => boolean) | any;
  executeOptions?: Omit<IAsyncActionRunOptions<A, R, T, N, S>, "key" | "cacheBreak">;
  watchLastValid?: boolean;
}

export type TRunWithPayload<R> = (func: (payload: R) => any) => any;

export interface IBaseObjResponseUse<A, R, T extends string, N, S extends IPullstateAllStores> {
  execute: (runOptions?: IAsyncActionRunOptions<A, R, T, N, S>) => TPullstateAsyncRunResponse<R, T, N>;
}

export interface IBaseObjResponseUseDefer<A, R, T extends string, N, S extends IPullstateAllStores> {
  execute: (args?: A, runOptions?: Omit<IAsyncActionRunOptions<A, R, T, N, S>, "key" | "cacheBreak">) => TPullstateAsyncRunResponse<R, T, N>;
  hasCached: (args?: A, options?: { successOnly?: boolean } & Omit<IAsyncActionGetCachedOptions<A, R, T, N, S>, "key">) => boolean;
  unwatchExecuted: () => void;
  useDebouncedExecution: (args: A, delay: number, options?: IUseDebouncedExecutionOptions<A, R, T, N, S>) => void;
  args: A;
  key: string;
}

export interface IBaseObjResponse<R, T extends string, N> {
  isLoading: boolean;
  isFinished: boolean;
  isUpdating: boolean;
  isStarted: boolean;
  // isSuccess: boolean;
  // isFailure: boolean;
  clearCached: () => void;
  updateCached: (updater: TUpdateFunction<R>, options?: IAsyncActionUpdateCachedOptions) => void;
  setCached: (result: TAsyncActionResult<R, T, N>, options?: IAsyncActionSetOrClearCachedValueOptions) => void;
  setCachedPayload: (payload: R, options?: IAsyncActionSetOrClearCachedValueOptions) => void;
  endTags: (T | EAsyncEndTags)[];
  renderPayload: TRunWithPayload<R>;
  message: string;
  raw: TPullstateAsyncWatchResponse<R, T, N>;
}

export interface IBaseObjSuccessResponse<R, T extends string, N> extends IBaseObjResponse<R, T, N> {
  payload: R;
  errorPayload: null;
  error: false;
  isSuccess: true;
  isFailure: false;
}

export interface IBaseObjErrorResponse<R, T extends string, N> extends IBaseObjResponse<R, T, N> {
  payload: null;
  errorPayload: N;
  error: true;
  isFailure: true;
  isSuccess: false;
}

export type TUseResponse<A = any, R = any, T extends string = string, N = any, S extends IPullstateAllStores = IPullstateAllStores> =
  (IBaseObjSuccessResponse<R, T, N>
    | IBaseObjErrorResponse<R, T, N>) & IBaseObjResponseUse<A, R, T, N, S>;

export type TUseDeferResponse<A = any, R = any, T extends string = string, N = any, S extends IPullstateAllStores = IPullstateAllStores> =
  (IBaseObjSuccessResponse<R, T, N>
    | IBaseObjErrorResponse<R, T, N>) & IBaseObjResponseUseDefer<A, R, T, N, S>;
