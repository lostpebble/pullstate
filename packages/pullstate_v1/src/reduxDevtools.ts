import { IPullstateAllStores } from "./PullstateCore";

interface IORegisterInDevtoolsOptions {
  namespace?: string;
}

export function registerInDevtools(stores: IPullstateAllStores, { namespace = "" }: IORegisterInDevtoolsOptions = {}) {
  const devToolsExtension = typeof window !== "undefined" ? (window as any)?.__REDUX_DEVTOOLS_EXTENSION__ : undefined;

  if (devToolsExtension) {
    for (const key of Object.keys(stores)) {
      const store = stores[key];

      const devTools = devToolsExtension.connect({ name: `${namespace}${key}` });
      devTools.init(store.getRawState());
      let ignoreNext = false;
      /*store.subscribe(
          (state) => {
            if (ignoreNext) {
              ignoreNext = false;
              return;
            }
            devTools.send("Change", state);
          },
          () => {}
        );*/
      store.subscribe(
        (s) => s,
        (watched) => {
          if (ignoreNext) {
            ignoreNext = false;
            return;
          }
          devTools.send("Change", watched);
        }
      );

      devTools.subscribe((message: { type: string; state: any }) => {
        if (message.type === "DISPATCH" && message.state) {
          ignoreNext = true;
          const parsed = JSON.parse(message.state);
          store.replace(parsed);
        }
      });
    }
  }
}
