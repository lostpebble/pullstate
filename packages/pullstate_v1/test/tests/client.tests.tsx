import ReactDOMServer from "react-dom/server";
import React from "react";
import { InjectStoreState, update, useStoreState } from "../../src/index";
import { TestUIStore } from "./testStores/TestUIStore";
const beautify = require("js-beautify").html;

const Counter = () => {
  const count = useStoreState(TestUIStore, s => s.count);

  return (
    <div>
      <h3>Count: {count}</h3>
      <button
        onClick={() =>
          update(TestUIStore, (s, o) => {
            s.count = o.count + 1;
          })
        }
      >
        PLUS
      </button>
    </div>
  );
};

const App = () => {
  return (
    <div>
      <h1>Some test</h1>
      <InjectStoreState store={TestUIStore} on={s => s.message}>
        {message => (
          <div>
            <h2>{message}</h2>
            <input
              onChange={e =>
                TestUIStore.update(s => {
                  s.message = e.target.value;
                })
              }
              value={message}
            />
          </div>
        )}
      </InjectStoreState>
      <InjectStoreState store={TestUIStore}>{uiStore => <h2>{uiStore.count}</h2>}</InjectStoreState>
      <Counter />
    </div>
  );
};

describe("Pullstate on client only", () => {
  const ReactApp = <App />;

  it("Should be able to render its data", () => {
    expect(beautify(ReactDOMServer.renderToString(ReactApp))).toMatchSnapshot();
  });
});
