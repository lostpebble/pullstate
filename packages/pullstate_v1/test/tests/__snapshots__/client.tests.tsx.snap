// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pullstate on client only Should be able to render its data 1`] = `
"<div data-reactroot=\\"\\">
    <h1>Some test</h1>
    <div>
        <h2>what what!</h2><input value=\\"what what!\\" />
    </div>
    <h2>5</h2>
    <div>
        <h3>Count:
            <!-- -->5</h3><button>PLUS</button>
    </div>
</div>"
`;
